enterprise/kernel/src/main/java/org/neo4j/kernel/impl/pagecache/WarmupAvailabilityListener.java

```diff
@@ -0,0 +1,118 @@
+import static org.neo4j.scheduler.JobScheduler.Groups.pageCacheIOHelper;
+
+class WarmupAvailabilityListener implements AvailabilityListener
+{
+    private final JobScheduler scheduler;
+    private final PageCacheWarmer pageCacheWarmer;
+    private final Config config;
+    private final Log log;
+    private final PageCacheWarmerMonitor monitor;
+    private volatile JobScheduler.JobHandle profileHandle;
+
+    WarmupAvailabilityListener( JobScheduler scheduler, PageCacheWarmer pageCacheWarmer,
+                                Config config, Log log, PageCacheWarmerMonitor monitor )
+    {
+        this.scheduler = scheduler;
+        this.pageCacheWarmer = pageCacheWarmer;
+        this.config = config;
+        this.log = log;
+        this.monitor = monitor;
+    }
+
+    @Override
+    public void available()
+    {
+        scheduler.schedule( pageCacheIOHelper, this::startWarmup );
+    }
+
+    private void startWarmup()
+    {
+        try
+        {
+            long start = System.nanoTime();
+            pageCacheWarmer.reheat().ifPresent( pagesLoaded ->
+            {
+                long elapsedMillis = NANOSECONDS.toMillis( System.nanoTime() - start );
+                monitor.warmupCompleted( elapsedMillis, pagesLoaded );
+                log.debug( "Active page cache warmup took " + Format.duration( elapsedMillis ) +
+                           " to load " + pagesLoaded + " pages." );
+            } );
+        }
+        catch ( Exception e )
+        {
+            log.debug( "Active page cache warmup failed, " +
+                       "so it may take longer for the cache to be populated with hot data.", e );
+        }
+
+        scheduleProfile();
+    }
+
+    private void scheduleProfile()
+    {
+        long frequencyMillis = config.get( GraphDatabaseSettings.pagecache_warmup_profiling_interval ).toMillis();
+        profileHandle = scheduler.scheduleRecurring(
+                pageCacheIOHelper, this::doProfile, frequencyMillis, TimeUnit.MILLISECONDS );
+    }
+
+    private void doProfile()
+    {
+        try
+        {
+            long start = System.nanoTime();
+            pageCacheWarmer.profile().ifPresent( pagesInMemory ->
+            {
+                long elapsedMillis = NANOSECONDS.toMillis( System.nanoTime() - start );
+                monitor.profileCompleted( elapsedMillis, pagesInMemory );
+                log.debug( "Profiled page cache in " + Format.duration( elapsedMillis ) +
+                           ", and found " + pagesInMemory + " pages in memory." );
+            } );
+        }
+        catch ( Exception e )
+        {
+            log.debug( "Page cache profiling failed, so no new profile of what data is hot or not was produced. " +
+                       "This may reduce the effectiveness of a future page cache warmup process.", e );
+        }
+    }
+
+    @Override
+    public void unavailable()
+    {
+        JobScheduler.JobHandle handle = profileHandle;
```

Comment:
I don't see the benefit of this. It does make it less obvious that it is the same object having `cancel` called and being set to null below.

Reply:
I could make the null-out use `compareAndSet`, but perhaps this exercise is overkill?

Reply:
Ah right, I get it now. Maybe an `AtomicReference` could be used, but as you say maybe it would be overkill.